### PR TITLE
New primitive hard block support titan flow

### PR DESCRIPTION
**Pull request by Srivat97**
Added support to process designs that contain new types of hard blocks.

#### Description

When new primitive hard blocks are included in a design, the Quartus synthesis tool does not correctly include the hard blocks
within the generated vqm netlists. Instead of directly instantiating the hard block within the netlist (which occurs for BRAM or DSP blocks), LUTs and DFFs have been added in place of the ports of each hard block within the design.

The feature added in this PR goes through netlists that contain hard blocks and remove the unnecessary LUTs and DFFs and in their place instantiate the corresponding hard block. The resulting blif file correctly includes all the hard blocks. 

#### Motivation and Context

Titan Flow currently does not support new types of hard blocks, so designers would have to manually add the hard blocks to the netlist. This feature automates the previous step and quickly allows designers to perform tests and evaluate architectures with new types of hard blocks.

#### How Has This Been Tested?

The following tests were performed:
- A custom design with a new type of hard block was used to verify the functionality of the new feature
- Sanitization and Valgrind tests were also performed
- The existing Titan23 circuits were used to verify that the new feature did not modify prior results

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [ x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x ] I have added tests to cover my changes
- [ x] All new and existing tests passed

### CODE REVIEW

**Comment by Srivat97** [comment]
@vaughnbetz I resolved most of the changes you had, but there were a few changes that I commented on. Let me know your thoughts on them.

File: utils/vqm2blif/src/base/hard_block_recog.h
Comment (vaughnbetz) [reply utils/vqm2blif/src/base/hard_block_recog.h]: OK, can leave then.

**Review by vaughnbetz — COMMENTED** [review]
Looks good Srivatsan.  This can be merged once CI goes green.
I added a couple of minor comment notes / typo that wasn't fixed, but those can be changed after the initial merge.
Also, you can reply to comments in a code review by typing in the "reply" box; that way the conversation will show up with your responses under your name (so it's a bit more clear than editing my notes).

**Review by vaughnbetz — APPROVED** [review]

**Comment by vaughnbetz** [comment]
There's one spurious QoR failure:
vtr_reg_nightly_test1/arithmetic_tasks/multless_consts...[Fail]
[Fail]
fixed_k6_frac_N8_22nm.xml/mult_057.v/common min_chan_width_route_time relative value 14.643312101910826 outside of range [0.1,14.0], above absolute threshold 2.0 and not equal to golden value: 1.57

But that isn't a major concern (min chan width route time bounces around) and it isn't related to your change. Merging.

**Pull request merged**
